src/OctoshiftCLI.Tests/AdoClientTests.cs

```diff
@@ -0,0 +1,1042 @@
+        }
+
+        [Fact]
+        public async Task PutAsync_Logs_The_Response_Status_Code_And_Content()
+        {
+            // Arrange
+            using var httpClient = new HttpClient(MockHttpHandlerForPut().Object);
+            var adoClient = new AdoClient(_loggerMock.Object, httpClient, PERSONAL_ACCESS_TOKEN);
+
+            // Act
+            await adoClient.PutAsync("http://example.com/resource", _rawRequestBody);
+
+            // Assert
+            _loggerMock.Verify(m => m.LogVerbose($"RESPONSE ({_httpResponse.StatusCode}): {EXPECTED_RESPONSE_CONTENT}"), Times.Once);
+        }
+
+        [Fact]
+        public async Task PutAsync_Throws_HttpRequestException_On_Non_Success_Response()
+        {
+            // Arrange
+            using var httpResponse = new HttpResponseMessage(HttpStatusCode.InternalServerError);
+            var handlerMock = MockHttpHandler(_ => true, httpResponse);
+
+            // Act
+            // Assert
+            await FluentActions
+                .Invoking(() =>
+                {
+                    using var httpClient = new HttpClient(handlerMock.Object);
+                    var adoClient = new AdoClient(_loggerMock.Object, httpClient, PERSONAL_ACCESS_TOKEN);
+                    return adoClient.PutAsync("http://example.com/resource", _rawRequestBody);
+                })
+                .Should()
+                .ThrowExactlyAsync<HttpRequestException>();
+        }
+
+        [Fact]
+        public async Task PatchAsync_Encodes_The_Url()
+        {
+            var handlerMock = MockHttpHandlerForPatch();
+            using var httpClient = new HttpClient(handlerMock.Object);
+            var adoClient = new AdoClient(_loggerMock.Object, httpClient, PERSONAL_ACCESS_TOKEN);
+
+            const string actualUrl = "http://example.com/param with space";
+            const string expectedUrl = "http://example.com/param%20with%20space";
+
+            // Act
+            await adoClient.PatchAsync(actualUrl, _rawRequestBody);
+
+            // Assert
+            handlerMock.Protected().Verify(
+                "SendAsync",
+                Times.Once(),
+                ItExpr.Is<HttpRequestMessage>(msg => msg.RequestUri.AbsoluteUri == expectedUrl),
+                ItExpr.IsAny<CancellationToken>());
+        }
+
+        [Fact]
+        public async Task PatchAsync_Applies_Retry_Delay()
+        {
+            // Arrange
+            using var firstHttpResponse = new HttpResponseMessage(HttpStatusCode.OK)
+            {
+                Headers = { RetryAfter = new RetryConditionHeaderValue(TimeSpan.FromSeconds(1)) },
+                Content = new StringContent("FIRST_RESPONSE")
+            };
+            using var secondHttpResponse = new HttpResponseMessage(HttpStatusCode.OK)
+            {
+                Content = new StringContent("SECOND_RESPONSE")
+            };
+            using var thridResponse = new HttpResponseMessage(HttpStatusCode.OK)
```

Comment:
```suggestion
            using var thirdResponse = new HttpResponseMessage(HttpStatusCode.OK)
```

Reply:
I'll stop here. Looks like this mistake is for each HTTP method (PUT, POST, GET, etc)